### DISCUSSION TITLE
Support forwarding ports from non-`localhost` hosts

### DISCUSSION THREAD

**Discussion by jarrodldavis** [community/community#9420] [Codespaces]
The other Visual Studio Code Remote extensions [support forwarding ports from non-`localhost` hosts](https://github.com/microsoft/vscode-remote-release/issues/3450) inside the container, so GitHub Codespaces should support it too. This is useful for exposing ports from Docker Compose services, such as a database, without having to use the usual hack of putting both containers on the same network (like is done [here](https://github.com/microsoft/vscode-dev-containers/blob/6e97392e224a3bcf392f3b5b1cd171697e4b4764/containers/python-3-postgres/.devcontainer/docker-compose.yml#L22-L23) and [here](https://github.com/microsoft/vscode-dev-containers/blob/6e97392e224a3bcf392f3b5b1cd171697e4b4764/containers/ruby-rails-postgres/.devcontainer/docker-compose.yml#L22-L23)).

I've tried doing this on Codespaces (even following [these testing directions](https://github.com/microsoft/vscode-remote-release/issues/5635) using [this repository](https://github.com/jarrodldavis/codespaces-port-forward-test)) and it doesn't work. `forwardPorts` entries in `devcontainer.json` appear to be ignored if they are a `host:port` string, and manually forwarding a port using a `host:port` string appears to silently drop the host and only forward the port from `localhost` in the container.

**Comment by shellerci** [comment]
If you checkout the [`devcontainer.json`](https://code.visualstudio.com/docs/remote/devcontainerjson-reference#_general-devcontainerjson-properties) for forwardPorts they've got the disclaimer : 

> ⚠️ Codespaces does not yet support the "`host:port`" variation of this property

Which I agree is less than optimal, although the "does not yet" wording gives me hope that this is on their todo list. 

In the interim, I've been using using [socat](https://www.redhat.com/sysadmin/getting-started-socat) as a workaround. 

Added this to my `Dockerfile`
```
RUN apt-get install socat
```

Added MySQL service to my  `docker-compose.yml`
```
  cs-mysql:
    image: mysql:5.7
    restart: unless-stopped
```

Added this to my forwardPorts in `devcontainer.json`
```
forwardPorts: ["3306"]
```
And the following to my `postCreate` script 
```
# socat TCP4-LISTEN:<PORT-TO-FORWARD>,reuseaddr,fork TCP:<SERVICE-NAME>:<PORT-TO-FORWARD> & 
socat TCP4-LISTEN:3306,reuseaddr,fork TCP:cs-mysql:3306 & 
```

**Reply by jarrodldavis** [reply]
Ah weird I must have missed that when reading the documentation before. I know it's still technically in beta but having Codespaces lag behind other Remote extensions like this is frustrating when trying to make a `devcontainer.json` setup that you want to work for both local Docker container usage and Codespaces. Hopefully the Codespaces team has time to work on closing this and other gaps soon. :)

And yeah, I also figured out that `socat` works as a workaround, but as you said it's less than ideal.

**Reply by longzheng** [reply]
In case anyone else was also looking for this, the disclaimer got moved to https://containers.dev/supporting#codespaces-specific-limitations

![image](https://github.com/user-attachments/assets/267dbc3f-3df7-4974-9915-55d7b5bf77c0)
